### ISSUE TITLE
IPv6 support

### ISSUE THREAD

**Issue by mwarning** [nodogsplash/nodogsplash#38]
Hi,

NDS does not support IPv6 - but it is the future.
We could port libhttpd, bu maybe it makes sense to use a different HTTP library?

**Comment by lynxis** [comment]
for both versions: The firewall handling is missing for ipv6, right?

**Comment by mwarning** [comment]
yes and yes :-)

**Comment by wanxewoj** [comment]
I created a pull request that fixes the IPv6 issues in `fw_iptables.c` causing nodogsplash to immediately shutdown again: https://github.com/nodogsplash/nodogsplash/pull/311

This is just another step towards IPv6 support.

**Comment by wanxewoj** [comment]
Another place I've been working on is `http_microhttpd.c`. A main challenge there is that function `arp_get()` is using `/proc/net/arp` for looking up the MAC address of the requester: https://github.com/nodogsplash/nodogsplash/blob/e735416f5166df4b5f12c77af268ed9061e943bd/src/http_microhttpd.c#L231

However, this approach (via procfs) only works for IPv4, as there is no equivalent procfs entry for IPv6 neighbours. I found a generic solution via a [NETLINK_ROUTE socket](https://en.wikipedia.org/wiki/Netlink#Netlink_protocols)  that works for both IPv4 and IPv6, but it adds a new library dependency: [libnl-route](https://www.infradead.org/~tgr/libnl/doc/api/group__neigh.html#gac0f72f54c8c5f5cacf0f1aae5807ae00).

**Therefore I would like to ask first if this is acceptable?** 
For instance OpenWrt 18.06 by default only ships `libnl-tiny` which is not sufficient, as it doesn't contain `libnl-route`. NDS would require `libnl-full` in addition to `libmircohttpd` if we decide to go this direction.

**Comment by mwarning** [comment]
@wanxewoj you ca use "ip neigh" to get the neighbour table. Add `-4` or `-6` if you want to have IPv4/IPv6 only.

**Comment by wanxewoj** [comment]
> you ca use "ip neigh" to get the neighbour table

Sure. But then you have to parse the text output. I'm not sure if this could differ from system to system or change over time? Do we really want to consider the output of `ip neigh` as a `stable API` for NDS?

**Comment by mwarning** [comment]
We currently parse text output anyway; it's not very efficient nor very stable. But it adds no dependencies and is easy to do. Also, future developers have it easier to understand the code.

**Comment by wanxewoj** [comment]
Well, then the easiest and most understandable solution would be to pipe `ip neigh show` with `grep` and `awk`:

```C++
int
get_mac(char mac[18], const char req_ip[])
{
	char msg[255] = {0};
	int rc;
	
	rc = execute_ret(msg, sizeof(msg) - 1, 
		"ip neigh show | grep %s | awk '{print $5}'", req_ip);	
	
	if (rc != 0) {
		return -1;
	} else {
		strncpy(mac, msg, 17);
		return 0;
	}
}
```
This works for IPv4 and IPv6, because `grep %s` only returns the given line and `awk` just filters the 5th column which is the MAC address in both cases (IPv4 and IPv6).

But can `grep` and `awk` assumed to be available on every OpenWrt system by default?

**Comment by mwarning** [comment]
grep and awk is on every OpenWrt system. But it is probably much faster to search for the string in C. grep and awk start entire new processes.

Here is some unfinished code that matches the line with the ip address:
```
int
get_mac(char mac[18], const char req_ip[])
{
	char line[255] = {0};
	char ip[64];
	int len;
	FILE *stream;

	// Extend search string by a space
	len = strlen(req_ip);
	memcpy(ip, req_ip, len);
	ip[len] = ' ';
	ip[len+1] = '\0';

	stream = popen("ip neigh show", "r");
	while (!feof(stream)) {
		if (fgets(line, sizeof(line) - 1, stream) != NULL) {
			if (0 == strncmp(req_ip, line, len)) {
				// line found
				printf("found\n");
				// TODO: need to extract mac from column 5 of the line
				break;
			}
		}
	}
	pclose(stream);
}
```

**Comment by mwarning** [comment]
That should do..
```C
int
get_mac(char mac[18], const char req_ip[])
{
	char line[255] = {0};
	char ip[64];
	FILE *stream;
	int len;
	int ret;

	ret = 0;
	len = strlen(req_ip);

	if ((len + 1) < sizeof(ip)) {
		// Extend search string by one space
		memcpy(ip, req_ip, len);
		ip[len] = ' ';
		ip[len+1] = '\0';

		stream = popen("ip neigh show", "r");
		while (!feof(stream)) {
			if (fgets(line, sizeof(line) - 1, stream) != NULL) {
				if (0 == strncmp(line, ip, len + 1)) {
					ret = (1 == sscanf(line, "%*s %*s %*s %*s %17[A-Fa-f0-9:] ", mac));
					break;
				}
			}
		}

		pclose(stream);
	}

	return ret;
}
```

**Comment by mwarning** [comment]
Applied: https://github.com/nodogsplash/nodogsplash/commit/53d5b16b1d30bebea16291599f2826033dc8d147

**Comment by wanxewoj** [comment]
Next issue is that [inet_ntop()](https://github.com/nodogsplash/nodogsplash/blob/176e58bd40751dbd498bccefb34145daf0e2f50e/src/http_microhttpd.c#L271) throws an `ENOSPC` error in the `AF_INET6` case:

```C++
case AF_INET6:
    if (inet_ntop(addrin6->sin6_family, &(addrin6->sin6_addr), ip_addr, sizeof(struct sockaddr_in6))) {
        return 0;
    }
```

The problem is that `sizeof(struct sockaddr_in6)` does not return the correct size for IPv6 addresses. I think `INET6_ADDRSTRLEN` should be used instead of `sizeof()`. However, this means that it is coincidence that the IPv4 case works right now, because it should be changed there as well!?

Here's my suggestion for discussion:
```C++
static int
get_client_ip(char ip_addr[INET6_ADDRSTRLEN+1], struct MHD_Connection *connection)
{
	...
	const struct sockaddr_in *addrin;
	const struct sockaddr_in6 *addrin6;

	...

	switch(client_addr->sa_family) {
	case AF_INET:
		if (inet_ntop(addrin->sin_family, &(addrin->sin_addr), ip_addr, INET_ADDRSTRLEN)) {
			return 0;
		}
		break;

	case AF_INET6:
		if (inet_ntop(addrin6->sin6_family, &(addrin6->sin6_addr), ip_addr, INET6_ADDRSTRLEN)) {
			return 0;
		}
		break;
	}

	return -1;
}
```

**Comment by wanxewoj** [comment]
One more comment/question. I noticed that `fw_iptables.c`, `http_microhttpd.c` and `util.c` always contain INET6_ADDRSTRLEN **+1** when char arrays are defined. However, according to [this discussion](https://stackoverflow.com/questions/39443413/why-is-inet6-addrstrlen-defined-as-46-in-c#39443536) _INET_ADDRSTRLEN_ and _INET6_ADDRSTRLEN_ are each already defined 1 byte bigger than the actual maximum address length to respect the terminating null byte. Therefore the `+1` in our code shouldn't be needed, should it?

Example:
```C++
static int
get_client_ip(char ip_addr[INET6_ADDRSTRLEN+1], struct MHD_Connection *connection)
{ ... }
```

**Comment by mwarning** [comment]
You are correct regarding inet_ntop and INET6_ADDRSTRLEN+1. Please make a pull request an reap the karma. :-)

**Comment by wanxewoj** [comment]
Next: There are many places in the code base, where a URL string is build like:
```C++
safe_asprintf(&denyaction, "http://%s:%d/%s/", config->gw_address, config->gw_port, config->denydir);
safe_asprintf(&authaction, "http://%s:%d/%s/", config->gw_address, config->gw_port, config->authdir);
safe_asprintf(&authtarget, "http://%s:%d/%s/?tok=%s&amp;redir=%s", config->gw_address, config->gw_port, config->authdir, client->token, redirect_url);
```

However, for IPv6 the IP address must always be wrapped in **square brackets**. I'm considering a helper function in `util.c`:

```C++
const char *
format_url(char *ip, unsigned int port) 
{
    char *proto = "http://";

    /* The formated url this function will return.
     * The maximum length will be:
     * - sizeof protocol string (e.g. 'http://')
     * - INET6_ADDRSTRLEN (46 byte, incl. terminating NULL byte)
     * - IP and port separator ':' (1 byte)
     * - sizeof port (up to 5 digits/byte, max '65535')
     * TODO: is it safe/good idea to define a static variable here? 
     **/
    static char url[sizeof(proto)+INET6_ADDRSTRLEN+1+5];

    /* test if ip is v4 or v6 */
    char buf[16];
    if (inet_pton(AF_INET, ip, buf)) {
        sprintf(url, "%s%s:%d", proto, ip, port);
    } else if (inet_pton(AF_INET6, ip, buf)) {
    	/* add square brackets around IPv6 address */
        sprintf(url, "%s[%s]:%d", proto, ip, port);
    }

    return url;
}
```

So the above mentioned lines can be easily changed to:
```C++
safe_asprintf(&denyaction, "%s/%s/", format_url(config->gw_address, config->gw_port), config->denydir);
safe_asprintf(&authaction, "%s/%s/", format_url(config->gw_address, config->gw_port), config->authdir);
safe_asprintf(&authtarget, "%s/%s/?tok=%s&amp;redir=%s", format_url(config->gw_address), config->gw_port, config->authdir, client->token, redirect_url);
```

Any comments on this approach?

**Comment by mwarning** [comment]
How about we rename `config->gw_address` to `config->gw_ip` and use `config->gw_address` to denote ip + port:
```
config->gw_address = "1.2.3.4:80"
config->gw_address = "1.2.3.4"
config->gw_port = 80
```

Then we can use:
```
safe_asprintf(&denyaction, "http://%s/%s/", config->gw_address, config->denydir);
```

**Comment by mwarning** [comment]
Also, `config->ip6` should say if the gateway address is an ipv4 or ipv6 address.
But maybe `config->gw_address` should be in a binary representation at some point (sockaddr_in/sockaddr_in6) that also encodes the protocol.

**Comment by wanxewoj** [comment]
> How about we rename config->gw_address...

> Also, config->ip6 should say if the gateway address is an ipv4 or ipv6 address...

This is simpler than a dedicated util function, but only if NDS will run in single-stack mode (IPv4 or IPv6, but not both at the same time).

**Comment by wanxewoj** [comment]
But it's probably easier to focus on single-stack mode (IPv6 only) first and once this works consider refactoring the code for dual-stack operation.

**Comment by mwarning** [comment]
We want to run NDS in dual-stack mode all the time if we can get ipv6 support to work as well as ipv4.
That means that the libmicrohttpd_cb method will have to deal with either ipv4 or ipv6 addresses, depending on the client.

The current approach is to improve ipv6 support (single stack) and to hide it behind config->ipv6. The main point of config->ipv6 was to drive ipv6 support forward and not to break ipv4.

**Comment by mwarning** [comment]
@wanxewoj it would be awesome if you can get dual-stack working. This might have some impact on the config. But we can probably keep it small and even backwards compatible.

**Comment by wanxewoj** [comment]
> How about we rename config->gw_address to config->gw_ip and use config->gw_address to denote ip + port

OK. I'll proceed with that approach and see how far we'll get.

**Comment by mwarning** [comment]
@bluewavenet this won't break FAS. :-)

**Comment by bluewavenet** [comment]
@mwarning 
I am all in favour of introducing ipv6 support, but now we are about to go beyond just a few minor changes to prepare the way.

We are at the stage where the potential exists to break many things including FAS. 
I do see all this as progress though, and it is good. 

@wanxewoj I admire your enthusiasm.

In my opinion, now is the time to create a new NDS Version 4 branch and confine any ipv6 development to that.

This will make it so much easier to just try things out without any consequences.

In the UK, where I am based, very few ISPs even give the option of ipv6 and if they do then usually most likely only in data centres. 

I know this will change / is changing slowly but for now it is essential that we do not break NDS ipv4. 
I will go as far to say that I will currently have some difficulty setting something up to test with ipv6 and I certainly don't have a requirement for it just yet.

I would love to be able to say in a couple of years or so that we can now drop ipv4, but I do not see that happening for quite some time longer than that.

**Comment by mwarning** [comment]
@bluewavenet I do not think you really need to test the ipv6 handling if you do not use ipv6. You let is disabled in NDS or disable ipv6 system wide on openwrt.

How about we look at the changes @wanxewoj might come up with and decide over the pull request if there need to be a development branch for now.

We won't enable ipv6 by default in the next release. That would be too hasty.

**Comment by mwarning** [comment]
ok :-)

**Comment by wanxewoj** [comment]
With the latest fixes NDS is running now with IPv6 and serving the splash page, status page. As well as authenticating the client via basic token authentication. However, the client has to explicitly open http://[NDS_IPV6]:2050 to trigger this process. 
`ip6tables` part is not complete yet, therefore clients are not automatically captured yet.

**Comment by wanxewoj** [comment]
The final piece seems to be finding an equivalent replacement for [the 6 NAT rules starting on line 510](https://github.com/nodogsplash/nodogsplash/blob/1a05c1c7ad0362b45a98e84703bd9f8178a1d473/src/fw_iptables.c#L510) `fw_iptables.c`.

IPv6 doesn't have the `nat` table, but it has the `mangle` table which also has a `PREROUTING` chain. However, the `-j DNAT` target is not available in IPv6. So line [524](https://github.com/nodogsplash/nodogsplash/blob/1a05c1c7ad0362b45a98e84703bd9f8178a1d473/src/fw_iptables.c#L524) is the critical one. The alternative `-j REDIRECT` target also doesn't exist for IPv6.

However, there is the [TPROXY](https://www.mjmwired.net/kernel/Documentation/networking/tproxy.txt) extension which provides similar functionality. It won't change the packet's destination like NAT, but it will mark the packets if a matching local socket is found. Via routing rules you can achieve that those packets are then routed to a local socket instead (e.g. libmicrohttpd 2050).

The only requirement seems to be that the socket was created with the `SOL_IP` and `IP_TRANSPARENT` flags. This doesn't seem to be supported by libmicrohttpd via the [MHD_OPTIONs](https://www.gnu.org/software/libmicrohttpd/manual/html_node/microhttpd_002dconst.html). But there is an `MHD_OPTION_LISTEN_SOCKET` flag which allows to pass an existing socket to MHD instead of it creating one. So this could be the solution, that we create the socket in `main.c`, setting the right flags and handing it over to MHD in [MHD_start_daemon()](https://github.com/nodogsplash/nodogsplash/blob/1a05c1c7ad0362b45a98e84703bd9f8178a1d473/src/main.c#L251).

Further resources:
https://people.netfilter.org/hidden/nfws/nfws-2008-tproxy_slides.pdf
https://github.com/kristrev/tproxy-example
https://wiki.squid-cache.org/Features/Tproxy4
https://serverfault.com/questions/927796/using-iptables-tproxy-instead-of-redirect

I could need some second and third opinion here! ;-)

**Comment by mwarning** [comment]
Maybe we can find a solution that does not need NAT. But I am not really good with iptables. Anyway, there is IPv6 NAT (I think the module is called nat66).

**Comment by bluewavenet** [comment]
ip6tables-mod-nat adds support for all the IPv4 nat commands afaik. No means of testing at the moment.

**Comment by bluewavenet** [comment]
@mwarning @wanxewoj 
I installed ip6tables-mod-nat and it allows me to add a DNAT from the command line. Can't test it though.

**Comment by mwarning** [comment]
@wanxewoj if it might help you, I can give you access to the branch. If that is ok for you.

**Comment by bawantha** [comment]
can I use Ipv6 address in NDSv4?

**Comment by bluewavenet** [comment]
@bawantha 
No.

> ip version 6 is not currently supported by NDS. It is not essential or advantageous to have in the short term but should be added at some time in the future.

https://nodogsplashdocs.readthedocs.io/en/stable/todo.html
